### PR TITLE
Support for displaying multimedia messages in Slack & Facebook

### PR DESCRIPTION
**Pull request by aarora91**
Support for displaying multimedia messages in both Slack and Facebook.
User specifies multimedia responses within the Conversation dialog node as follows:

```json
{
  "output": {
    "generic":[
      {
        "response_type": "text",
        "text": "Here are your nearest stores."
      },
      {
        "response_type": "image",
        "source": "http://..."
      },
      {
        "response_type": "option",
        "options": [
          {
            "label": "Location 1",
            "value:" "Location 1"
          },
          {
            "label": "Location 2",
            "value:" "Location 2"
          },
          {
            "label": "Location 3",
            "value:" "Location 3"
          }
        ]
      }
    ]
  }
}
```

All elements in the `generic` array are normalized to channel-specific format by the `normalize-conversation-for-channel` Cloud Function.

*Note: Multiple responses are grouped together in a single message for Slack but sent as several, individual replies in the case of Facebook.*

### CODE REVIEW

File: channels/facebook/receive/index.js [hunk 1]
```diff
@@ -287,6 +287,8 @@ function verifyFacebookSignatureHeader(params, auth) {
     .update(buffer, 'utf-8')
     .digest('hex');
 
+  console.log(`Calculated: ${calculatedHash}`);
+  console.log(`Expected: ${expectedHash}`);
```
Comment (dgterry) [inline channels/facebook/receive/index.js]: probably want to remove these

File: sample_workspace.json [hunk 2]
```diff
@@ -0,0 +1 @@
...:11:26.651Z"},{"text":"Did we reached restaurant","created":"2017-12-28T23:11:26.651Z","updated":"2017-12-28T23:11:26.651Z"},{"text":"direction to pharmacy","created":"2017-12-28T23:11:26.651Z","updated":"2017-12-28T23:11:26.651Z"},{"text":"Direction to San Diego Airport","created":"2017-12-28T23:11:26.651Z","updated":"2017-12-28T23:11:26.651Z"},{"text":"direction to santa monica","created":"2017-12-28T23:11:26.651Z","updated":"2017-12-28T23:11:26.651Z"},{"text":"Doctors nearby","created":"2017-12-28T23:11:26.651Z","updated":"2017-12-28T23:11:26.651Z"},{"text":"do you know any chinese restaurants around","created":"2017-12-28T23:11:26.651Z","updated":"2017-12-28T23:11:26.651Z"},{"text":"do you know a restaurant in new york city","created":"2017-12-28T23:11:26.651Z","updated":"2017-12-28T23:11:26.651Z"},{"text":"do you know a restaurant in new york city?","created":"2017-12-28T23:11:26.651Z","updated":"2017-12-28T23:11:26.651Z"},{"text":"do you know if there's any restaurant around here?","created":"2017-12-28T23:11:26.651Z","updated":"2017-12-28T23:11:26.651Z"},{"text":"Do you know pizza?","created":"2017-12-28T23:11:26.651Z","updated":"2017-12-28T23:11:26.651Z"},{"text":"do you kn
```
Comment (dgterry) [inline sample_workspace.json]: can we remove this file

File: channels/facebook/multiple_post/index.js [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+ * Recursively calls postMessage until all Conversation messages have been sent to the channel.
+ *
+ * @param sequenceName name of channel's postsequence sequence
+ * @param params the channel specific params to call the channel's api
+ * @param responses an array to store the response in case we have to build it over multiple posts
+ * @param size the number of messages to send to the channel
+ * @param index the array index of the message that needs to be sent
+ * @returns {*}
+ */
+function postMessage(sequenceName, params, responses, size, index) {
```
Comment (arrraichu) [inline channels/facebook/multiple_post/index.js]: here's a common way i've seen that shorthands and/or optimizes recursive promise calls (taken from [here](https://stackoverflow.com/questions/24586110/resolve-promises-one-after-another-i-e-in-sequence)):
```javascript
const sequence = Promise.resolve();

return facebook.messages
  .reduce((seq, message) => {
    seq
      .then(result => {
        return performAction(message);
      })
      .catch(error => {
        throw error;
      })
  }, sequence);
```
The Promise.resolve() `sequence` is the start of the promise chain, and every the chain resolves, the reducer will return and pass along the next promise to be called. This should make it easier to read, without worrying about array indices and the when-to-break-out-of-recursive-function business.

File: channels/facebook/multiple_post/index.js [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+}
+
+/**
+ * Uses the openwhisk npm module to invoke the channel's post action.
+ * @param actionName
+ * @param params
+ * @returns {boolean}
+ */
+function invokeAction(actionName, params) {
+  const ow = openwhisk();
```
Comment (arrraichu) [inline channels/facebook/multiple_post/index.js]: don't let this be called and reinitialized multiple times.

File: channels/facebook/multiple_post/index.js (hunk 3 above)
Comment (dgterry) [reply channels/facebook/multiple_post/index.js]: appreciate you teaching me this, I'm not going to hold up this current merge to develop as I don't want to slow things down but I hope to try this out in a future PR.

**Pull request merged**